K2Field.K2NE.ServiceBroker/ServiceObjects/CompressToZip/FilesToZip.cs

```diff
@@ -0,0 +1,128 @@
+            mFilesToZipSmartObject.InputProperties.Add(Constants.SOProperties.FilesToZip.FileName);
+            mFilesToZipSmartObject.Validation.RequiredProperties.Add(Constants.SOProperties.FilesToZip.FileName);
+            mFilesToZipSmartObject.ReturnProperties.Add(Constants.SOProperties.FilesToZip.ZipFile);
+
+            so.Methods.Add(mFilesToZipSmartObject);
+
+            soList.Add(so);
+
+            return soList;
+        }
+
+        public override void Execute()
+        {
+            switch (base.ServiceBroker.Service.ServiceObjects[0].Methods[0].Name)
+            {
+                case Constants.Methods.FilesToZip.FilesToZipMethod:
+                    FilesToZipMethod();
+                    break;
+            }
+        }
+
+
+        private void FilesToZipMethod()
+        {
+            ServiceObject serviceObject = ServiceBroker.Service.ServiceObjects[0];
+            serviceObject.Properties.InitResultTable();
+            DataTable results = ServiceBroker.ServicePackage.ResultTable;
+            string fileName = GetStringProperty(Constants.SOProperties.FilesToZip.FileName, true);
+            string query = GetStringProperty(Constants.SOProperties.FilesToZip.ADOSMOQuery, true);
+
+            DataTable SOQueryResult = new DataTable();
+
+            using (SOConnection connection = new SOConnection(base.BaseAPIConnectionString))
+            using (SOCommand command = new SOCommand(query, connection))
+            using (SODataAdapter adapter = new SODataAdapter(command))
+            {
+                connection.DirectExecution = true;
+                adapter.Fill(SOQueryResult);
+            }
+            string xmlZipFile = string.Empty;
+
+            if (SOQueryResult.Rows.Count > 0)
+            {
+                using (var memoryStream = new MemoryStream())
+                {
+                    using (var archive = new ZipArchive(memoryStream, ZipArchiveMode.Create, true))
+                    {
+                        foreach (DataRow row in SOQueryResult.Rows)
+                        {
+                            XElement fileXml = XElement.Parse(row[0].ToString());
+                            
+                            var demoFile = archive.CreateEntry(fileXml.Element("name").Value);
+
+                            using (var entryStream = demoFile.Open())
+                            using (var b = new BinaryWriter(entryStream))
+                            {
+                                b.Write(Convert.FromBase64String(fileXml.Element("content").Value));
+                            }
+                        }
+                    }
+                    
+                    string content = Convert.ToBase64String(memoryStream.ToArray());
+                    xmlZipFile = string.Format("<file><name>{0}.zip</name><content>{1}</content></file>", fileName, content);                    
```

Comment:
Consider using the ServiceSDK to create the XML. Creating it manually could introduce a [XML reserved character](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-reserved-xml-characters-in-project-files?view=vs-2019#use-reserved-characters).

For example:
```
var zipFile = serviceObject.Properties[Constants.SOProperties.FilesToZip.ZipFile] as FileProperty;
zipFile.FileName = fileName;
zipFile.Content = content;
xmlZipFile = zipFile.Value;
```